modules/core/02-client/keeper/extended_keeper.go

```diff
@@ -57,242 +76,85 @@ func (ek *ExtendedKeeper) SetHooks(ch ClientHooks) *ExtendedKeeper {
-	// if previous consensus state is not before current consensus state, freeze the client and return.
-	if prevOk && !prevCons.Timestamp.Before(consState.Timestamp) {
-		return types.ErrHeaderNonMonotonicTimestamp
-	}
-	// if next consensus state exists, check consensus state time is less than next consensus state time
-	// if next consensus state is not after current consensus state, freeze the client and return.
-	if nextOk && !nextCons.Timestamp.After(consState.Timestamp) {
-		return types.ErrHeaderNonMonotonicTimestamp
-	}
+	headerInfo := k.GetHeaderInfo(ctx, clientMsg)
 
-	return nil
-}
+	// found misbehaviour and it was not an header, freeze client
+	if foundMisbehaviour && headerInfo == nil {
+		clientState.UpdateStateOnMisbehaviour(ctx, k.cdc, clientStore, clientMsg)
 
-// checkValidity checks if the Tendermint header is valid.
-// CONTRACT: consState.Height == header.TrustedHeight
-// (copied from https://github.com/cosmos/ibc-go/blob/v5.0.0/modules/light-clients/07-tendermint/types/update.go#L168-L248)
-func checkValidity(
-	clientState *ibctmtypes.ClientState, consState *ibctmtypes.ConsensusState,
-	header *ibctmtypes.Header, currentTimestamp time.Time,
-) error {
-	if err := checkTrustedHeader(header, consState); err != nil {
-		return err
-	}
+		k.Logger(ctx).Info("client frozen due to misbehaviour", "client-id", clientID)
 
-	// UpdateClient only accepts updates with a header at the same revision
-	// as the trusted consensus state
-	if header.GetHeight().GetRevisionNumber() != header.TrustedHeight.RevisionNumber {
-		return sdkerrors.Wrapf(
-			ibctmtypes.ErrInvalidHeaderHeight,
-			"header height revision %d does not match trusted header revision %d",
-			header.GetHeight().GetRevisionNumber(), header.TrustedHeight.RevisionNumber,
+		defer telemetry.IncrCounterWithLabels(
+			[]string{"ibc", "client", "misbehaviour"},
+			1,
+			[]metrics.Label{
+				telemetry.NewLabel(types.LabelClientType, clientState.ClientType()),
+				telemetry.NewLabel(types.LabelClientID, clientID),
+				telemetry.NewLabel(types.LabelMsgType, "update"),
+			},
 		)
-	}
 
-	tmTrustedValidators, err := tmtypes.ValidatorSetFromProto(header.TrustedValidators)
-	if err != nil {
-		return sdkerrors.Wrap(err, "trusted validator set in not tendermint validator set type")
-	}
+		EmitSubmitMisbehaviourEvent(ctx, clientID, clientState)
 
-	tmSignedHeader, err := tmtypes.SignedHeaderFromProto(header.SignedHeader)
-	if err != nil {
-		return sdkerrors.Wrap(err, "signed header in not tendermint signed header type")
+		return nil
+	} else if foundMisbehaviour && headerInfo != nil {
+		// found misbehaviour and it was an header, this is most probably means
+		// conflicting headers misbehaviour.
+		ctx.Logger().Debug("received a header that has QC but is on a fork")
+		txHash := tmhash.Sum(ctx.TxBytes())
+		k.AfterHeaderWithValidCommit(ctx, txHash, headerInfo, true)
```

Comment:
Here `AfterHeaderWithValidCommit` only cares about Tendermint light client headers since the solo machine client may correspond to a solo machine that does not have consensus at all, e.g., [an account with a key pair](https://github.com/cosmos/ibc/blob/main/spec/client/ics-006-solo-machine-client/README.md). So probably we need to either check if `headerInfo` here corresponds to a `ibctmtypes.Header`, or make `GetHeaderInfo` return `nil` if the given header is `ibcsltypes.Header`.

Reply:
yea you are probably right, it does not make much sense for thius lc to be time stamped in any way. I think I will just return `nil` from `GetHeaderInfo` in that case, as then whole `UpdateClient` logic is basically the same as standard keeper `UpdateClient`

Reply:
yep sounds cool